// ng-lazyload-image/src/lazyload-image.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
require("rxjs/add/operator/let");
var core_1 = require("@angular/core");
var scroll_listener_1 = require("./scroll-listener");
var lazyload_image_1 = require("./lazyload-image");
var target = typeof window !== 'undefined' ? window : undefined;
var LazyLoadImageDirective = (function () {
    function LazyLoadImageDirective(el, ngZone) {
        this.scrollTarget = target;
        this.elementRef = el;
        this.ngZone = ngZone;
    }
    LazyLoadImageDirective.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.ngZone.runOutsideAngular(function () {
            if (_this.scrollObservable) {
                _this.scrollSubscription = _this.scrollObservable
                    .startWith('')
                    .let(lazyload_image_1.lazyLoadImage(_this.elementRef.nativeElement, _this.lazyImage, _this.defaultImage, _this.errorImage, _this.offset))
                    .subscribe(function () { });
            }
            else {
                _this.scrollSubscription = scroll_listener_1.getScrollListener(_this.scrollTarget)
                    .let(lazyload_image_1.lazyLoadImage(_this.elementRef.nativeElement, _this.lazyImage, _this.defaultImage, _this.errorImage, _this.offset))
                    .subscribe(function () { });
            }
        });
    };
    LazyLoadImageDirective.prototype.ngOnDestroy = function () {
        [this.scrollSubscription]
            .filter(function (subscription) { return subscription && !subscription.isUnsubscribed; })
            .forEach(function (subscription) { return subscription.unsubscribe(); });
    };
    return LazyLoadImageDirective;
}());
__decorate([
    core_1.Input('lazyLoad'),
    __metadata("design:type", Object)
], LazyLoadImageDirective.prototype, "lazyImage", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", String)
], LazyLoadImageDirective.prototype, "defaultImage", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", String)
], LazyLoadImageDirective.prototype, "errorImage", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], LazyLoadImageDirective.prototype, "scrollTarget", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], LazyLoadImageDirective.prototype, "scrollObservable", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Number)
], LazyLoadImageDirective.prototype, "offset", void 0);
LazyLoadImageDirective = __decorate([
    core_1.Directive({
        selector: '[lazyLoad]'
    }),
    __metadata("design:paramtypes", [core_1.ElementRef, core_1.NgZone])
], LazyLoadImageDirective);
exports.LazyLoadImageDirective = LazyLoadImageDirective;
